File: api/internal/infrastructure/logging/logger.go
package logging

import (
	"log/slog"
	"os"
)

var (
	// Logger is the global structured logger instance
	Logger *slog.Logger
)

// Init initializes the structured logger
// In production, use JSON handler. In development, use text handler.
func Init() {
	env := os.Getenv("ENV")
	if env == "production" || env == "prod" {
		// JSON handler for production (structured, searchable logs)
		opts := &slog.HandlerOptions{
			Level: slog.LevelInfo,
		}
		Logger = slog.New(slog.NewJSONHandler(os.Stdout, opts))
	} else {
		// Text handler for development (human-readable)
		opts := &slog.HandlerOptions{
			Level: slog.LevelDebug,
		}
		Logger = slog.New(slog.NewTextHandler(os.Stdout, opts))
	}
}

// GetLogger returns the global logger instance
// If not initialized, initializes with default settings
func GetLogger() *slog.Logger {
	if Logger == nil {
		Init()
	}
	return Logger
}

File: web/src/shared/lib/logging/logger.ts
/**
 * Structured logging using pino
 * Provides structured, searchable logs with appropriate log levels
 */

import pino from 'pino';

// Determine if we're in production
const isProduction = import.meta.env.PROD || import.meta.env.MODE === 'production';

// Configure logger based on environment
const loggerConfig: pino.LoggerOptions = {
	level: isProduction ? 'info' : 'debug',
	...(isProduction
		? {
				// Production: structured JSON output
				formatters: {
					level: (label) => {
						return { level: label };
					},
				},
			}
		: {
				// Development: human-readable output
				transport: {
					target: 'pino-pretty',
					options: {
						colorize: true,
						translateTime: 'HH:MM:ss Z',
						ignore: 'pid,hostname',
					},
				},
			}),
};

// Create and export logger instance
export const logger = pino(loggerConfig);

// Export convenience methods
export const log = {
	debug: (msg: string, ...args: unknown[]) => logger.debug(args.length > 0 ? { ...args } : {}, msg),
	info: (msg: string, ...args: unknown[]) => logger.info(args.length > 0 ? { ...args } : {}, msg),
	warn: (msg: string, ...args: unknown[]) => logger.warn(args.length > 0 ? { ...args } : {}, msg),
	error: (msg: string, error?: Error | unknown, ...args: unknown[]) => {
		if (error instanceof Error) {
			logger.error({ err: error, ...args }, msg);
		} else {
			logger.error(args.length > 0 ? { ...args, error } : { error }, msg);
		}
	},
};

export default logger;
